src/routes/(blog-article)/containerizing-rust-applications-best-practices/+page.md

```diff
@@ -0,0 +1,1156 @@
+
+<Callout type="info">
+
+The Linux Standard Base Core Specification defines three different ranges for
+user IDs, the value from 0 to 99, 100 to 999, and 1000 and above. The first two
+ranges are reserved for system users. All the regular users start from 1000.
+It's also important to note that the user ID 0 is reserved for the root user and
+Linux kernel uses the user id and group id to handle the permissions. The host
+machine and the docker container might have the same user ID but different usernames.
+
+</Callout>
+
+If you run that container, `appuser` will have the same permissions ad the user `1001`
+in the host system, whatever it's the name of tha user in the host machine.
+
+### Use The `docker run --user` Argument
+
+You can also overwrite the user running the container with the argument `--user`:
+
+<CodeBlock lang="console">
+
+```console
+docker run --user www-data --rm -it docker-rust-app-running-with-root whoami
+www-data
+```
+
+</CodeBlock>
+
+In this example, even though the image is execute as root by default, it will be
+executed as the `www-data` user.
+
+Notice you can even use a non-existing user in both the host and the docker image.
+
+<CodeBlock lang="console">
+
+```console
+$ docker run --user 1001 --rm -it docker-rust-app-running-with-root bash
+I have no name!@895b0f6a3dbb:/app$ 
+```
+
+</CodeBlock>
+
+All this solutions work but they all have a drawback: you need to know the user
+ID at build time (when you build the docker image). You usually want to run the
+container in different environments and sometimes you want to use a different
+user ID for each environment. For example:
+
+- For development: If you are using Ubuntu, your user ID is probably `1000`. When
+you run the container locally you want to run it using that ID, so that you don't
+have any problems with permissions.
+- For CI: The servers you are using for continuous integration (for instance, GitHub runners)
+might use an specif user. You could use some cache folders and maybe you need to
+use the same user ID as the CI server.
+- Production: You could create an specific user for your application and use that
+user ID.
+
+With the proposed solutions you would need to rebuild the docker image so that the
+user ID inside the container is the same as the host user id.
+
+### Create The User At Runtime
+
+There is al alternative to the previous solutions that make it possible to **run the
+container with different user IDs without rebuilding the image**.
+
+<CodeBlock lang="dockerfile">
+
+```dockerfile
+
+## Compile su-exec
+FROM docker.io/library/gcc:bookworm as gcc
+COPY ./contrib/dev-tools/su-exec/ /usr/local/src/su-exec/
+RUN cc -Wall -Werror -g /usr/local/src/su-exec/su-exec.c -o /usr/local/bin/su-exec; chmod +x /usr/local/bin/su-exec
+
+## Application
+FROM rust:1.73 as builder
+WORKDIR /app
+RUN cargo init
+RUN cargo build --release
+CMD ["./target/release/app"]
+
+## Runtime
+FROM gcr.io/distroless/cc-debian12:debug as runtime
```

Comment:
We use the :debug vairant of the distroless cc, as it includes the busybox binary, giving us a shell; that is needed to run the entry-script.

Reply:
Added.